File: prob2.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
//TC:O(n)
//SC:O(maxdepth)
class Solution {
    public void recoverTree(TreeNode root) 
    {
        
        Stack<TreeNode> stk = new Stack<>();
        TreeNode prev = null;
        TreeNode left = null;
        TreeNode middle=null;
        TreeNode right = null;
        while(root!=null ||!(stk.isEmpty()) )
        {
            while(root!=null)
            {
               stk.push(root); 
               root=root.left; 
            }
            
            TreeNode p =null;
             if(!stk.isEmpty())
              p = stk.pop();
            
            if(prev!=null && prev.val>p.val)
            {
             if(left==null)
             {
               left = prev;
               middle=p;
             }
             else
              right=p;
                
            }
              prev=p; 
              root=p.right;
  
        }
        
        if(right!=null) 
        {
          int temp = left.val;
          left.val=right.val;
          right.val =temp;  
        }
        
        else
        {
            int temp = left.val; 
            left.val = middle.val;
            middle.val = temp;
        }
        
    }
    
    
    
    
    
    
    
}

File: prob2.py
# Time Complexity :O(n)
# Space Complexity :O(h)
# Leet Code: Yes

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def recoverTree(self, root: Optional[TreeNode]) -> None:
        """
        Do not return anything, modify root in-place instead.
        """
        self.prev = self.first = self.second = None
        self.inorderTraversal(root)
        self.first.val, self.second.val = self.second.val, self.first.val
    
    def inorderTraversal(self, root):
        if root is None:
            return
        
        self.inorderTraversal(root.left)
        if self.prev is not None and root.val < self.prev.val:
            if self.first is None:
                self.first = self.prev
                self.second = root
            else:
                self.second = root
        self.prev = root
        self.inorderTraversal(root.right)
        